jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/ContextHandlerDeployer.java

```
@@ -0,0 +1,64 @@
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.deploy;
+
+import org.eclipse.jetty.server.handler.ContextHandler;
+
+/**
+ * Deployer for ContextHandlers
+ */
+public interface ContextHandlerDeployer
+{
+    /**
+     * Add a ContextHandler to the graph and then move it a deployed state.
+     *
+     * @param contextHandler the ContextHandler to deploy.
+     */
+    void deploy(ContextHandler contextHandler);
+
+    /**
+     * Move a ContextHandler int the graph to an undeployed state, and then remove it from the graph.
+     *
+     * @param contextHandler the ContextHandler to undeploy.
+     */
+    void undeploy(ContextHandler contextHandler);
+
+    /**
+     * Add a ContextHandler int the graph but perform no actions on it, leaving it in undeployed state.
+     *
+     * @param contextHandler the ContextHandler to add to the graph.
+     */
+    default void addUndeployed(ContextHandler contextHandler)
+    {
+    }
+
+    /**
+     * Advanced usage, move a ContextHandler through the DeploymentManager graph by name.
+     *
+     * @param contextHandler the ContextHandler to move
+     * @param goalName the goal graph node by name
+     */
+    default void move(ContextHandler contextHandler, String goalName)
+    {
+    }
+
+    /**
+     * Report a failure to the ContextHandlerDeployer to fail startup.
+     *
+     * @param cause the cause of the failure
+     */
+    default void reportStartupFailure(Throwable cause)
```

Why is this method on this interface?    Is it used by the scanner to report errors that it detects?  If so, more javadoc is required

---

It is documented.
Not sure what else you need to see here?

> Report a failure to the ContextHandlerDeployer to fail startup.

How about ...

If you need to report a failure during startup, use this API.
A failure during startup will fail the DeploymentManager startup as well.
This is important for failures that need to prevent successful startup.

---

Who calls this? Why is it needed? What is the Deployer meant to do with this? 

I'm not seeing why this is part of the abstract interface. If a "provider" has an error, it simply never calls add. If the Deployer had an error, well it is there Deployer implementation and doesn't need an abstract method to call itself?